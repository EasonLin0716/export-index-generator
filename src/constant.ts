export const supportExtensions = ['ts', 'tsx', 'js', 'jsx', 'mjs']
