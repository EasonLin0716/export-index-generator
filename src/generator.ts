import { ExportModule } from './types'
import type ts from 'typescript'
import { supportExtensions } from './constant'
import { analyzeExports, getExternalExports } from './analyze'
import { getSingleFileProgram } from './parser'
import { basename, extname } from 'path'

const removeNoNeedExtension = (path: string): string => {
  for (const ext of supportExtensions) {
    const result = path.replace(new RegExp(`\\.${ext}$`, 'g'), '')
    if (result !== path) return result
  }

  return path
}

const generateExports = (tsModule: typeof ts, path: string, exports: ExportModule[]): ts.ExportDeclaration[] => {
  const result: ts.ExportDeclaration[] = []
  const isTypeOnlyExports = exports.filter((item) => item.isTypeOnly)
  const otherExports = exports.filter((item) => !item.isTypeOnly)

  const createExportSpecifier = (item: ExportModule) => {
    const propertyName = item.default ? tsModule.createIdentifier('default') : undefined

    const [major, minor] = tsModule.version.split('.')
    if ((Number(major) === 4 && Number(minor) >= 5) || Number(major) > 4) {
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-expect-error
      return tsModule.createExportSpecifier(undefined, propertyName, tsModule.createIdentifier(item.name))
    }
    return tsModule.createExportSpecifier(propertyName, tsModule.createIdentifier(item.name))
  }

  const createExportDeclaration = (input: ExportModule[], isOnlyType: boolean) =>
    tsModule.createExportDeclaration(
      undefined,
      undefined,
      tsModule.createNamedExports(
        input.map(createExportSpecifier)
      ),
      tsModule.createStringLiteral(`./${removeNoNeedExtension(path)}`),
      isOnlyType
    )

  if (isTypeOnlyExports.length > 0) {
    result.push(createExportDeclaration(isTypeOnlyExports, true))
  }

  if (otherExports.length > 0) {
    result.push(createExportDeclaration(otherExports, false))
  }

  return result
}

export interface File {
  path: string
  content: string
}

export const generate = (tsModule: typeof ts, files: File[]): { content: string, scriptKind: ts.ScriptKind } => {
  const scriptKind = (() => {
    if (files.some((file) => extname(file.path) === '.tsx')) return tsModule.ScriptKind.TSX
    if (files.some((file) => extname(file.path) === '.ts')) return tsModule.ScriptKind.TS
    if (files.some((file) => extname(file.path) === '.jsx')) return tsModule.ScriptKind.JSX
    return tsModule.ScriptKind.JS
  })()

  const sourceFile = tsModule.createSourceFile(
    'index.ts',
    '',
    tsModule.ScriptTarget.ESNext,
    true,
    scriptKind
  )

  const statements = files.reduce(
    (result, file) => {
      // External
      if (!supportExtensions.some((ext) => file.path.endsWith(ext))) return result.concat(generateExports(tsModule, file.path, getExternalExports(file.path)))

      // Script
      const exports = analyzeExports(tsModule, getSingleFileProgram(tsModule, basename(file.path), file.content))
      return result.concat(generateExports(tsModule, file.path, exports))
    },
    [] as ts.ExportDeclaration[]
  )

  const printer = tsModule.createPrinter()

  return {
    content: printer.printFile(tsModule.updateSourceFileNode(sourceFile, statements)).replace(/;/g, '').replace(/"/g, '\''),
    scriptKind
  }
}
