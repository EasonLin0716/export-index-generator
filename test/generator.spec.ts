import { generate } from '../src/generator'
import ts from 'typescript'
import { supportExtensions } from '../src/constant'
import fg from 'fast-glob'
import { resolve } from 'path'
import { readFileSync } from 'fs'

const folders = ['fixture/ex1']

describe('test generator', () => {
  it('generate []', () => {
    expect(generate(ts, [])).toMatchSnapshot()
  })

  for (const folder of folders) {
    it(`generate ${folder}`, () => {
      const cwd = resolve(__dirname, folder)
      const files =
        fg
          .sync(`**/*.{${supportExtensions.join(',')}}`, { cwd }).filter((file) => !/^index\.[jt]{1}sx{0,1}$/.test(file))
          .map((path) => ({ path, content: readFileSync(resolve(cwd, path), { encoding: 'utf8' }) }))

      expect(generate(ts, files)).toMatchSnapshot()
    })
  }
})
