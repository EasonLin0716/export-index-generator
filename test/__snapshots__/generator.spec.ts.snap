// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test generator generate [] 1`] = `
Object {
  "content": "",
  "scriptKind": 1,
}
`;

exports[`test generator generate fixture/ex1 1`] = `
Object {
  "content": "export { Test5, Test6, default as Test7 } from './class'
export { Test16, Test17 } from './enum'
export type { Test13, default as Test15 } from './export'
export { Test14 } from './export'
export { test8, test9, test10 } from './function'
export type { Test2, Test3, default as Test4 } from './interface'
export type { Test11, Test12 } from './type'
",
  "scriptKind": 3,
}
`;
